nomad/job_endpoint_hooks.go

```diff
@@ -415,6 +421,17 @@ func (v *jobValidate) validateVaultIdentity(t *structs.Task) ([]error, error) {
 		return warnings, nil
 	}
 
+	// Only the default cluster is allowed to use the legacy flow. Using a
+	// non-default cluster requires an identity to be provided either in the
+	// task or as a default in the agent config.
```

Comment:
It's not clear to me how this conditional expresses what's in this comment. We're only checking for the non-default cluster and we'll always hit this unless we've hit an earlier set of errors.

Reply:
Most of the time we expect a task that uses Vault to have the expected identity, either provided directly in the task or via the agent config, so they end up here:
https://github.com/hashicorp/nomad/blob/142510b99c530c57a9cc9286678581b0426989fc/nomad/job_endpoint_hooks.go#L411-L422

If no identity is available, then we fall through the rest of the code in this function. Previously the code assumed that the legacy flow would be used, so there was only this:
https://github.com/hashicorp/nomad/blob/142510b99c530c57a9cc9286678581b0426989fc/nomad/job_endpoint_hooks.go#L435-L439

But non-default clusters are not allowed to use the legacy flow, so we need to check for this before proceeding with legacy flow validation.

I can clarify the comment on this.

Reply:
Ah, I got it now. I was misreading where the `hasTaskWID || hasDefaultWID` conditional always returns as though it only returned on a warning.

Reply:
Yeah, it's a bit confusing. I flipped the logic in https://github.com/hashicorp/nomad/commit/ace07aa14843223106167becba1f8d3462cc4b0a to look for problems first instead of returning early in the happy path. Is it easier to follow like this?

https://github.com/hashicorp/nomad/blob/ace07aa14843223106167becba1f8d3462cc4b0a/nomad/job_endpoint_hooks.go#L387-L434

Reply:
Oh that's nice. :+1:

Reply:
Cool! I ported that commit to this PR.